### PR TITLE
fix: move RoleManager to model and remove Invalid field in AddMatchingFunc

### PR DESCRIPTION
**Pull request by dovics**
fix: #501

### CODE REVIEW

File: model/model.go [hunk 1]
```diff
@@ -21,11 +21,16 @@ import (
...
+	"github.com/casbin/casbin/v2/rbac"
+	defaultrolemanager "github.com/casbin/casbin/v2/rbac/default-role-manager"
 	"github.com/casbin/casbin/v2/util"
 )
 
 // Model represents the whole access control model.
-type Model map[string]AssertionMap
+type Model struct {
+	Map map[string]AssertionMap
+	RM  rbac.RoleManager
```
[the remark this answers is not in the document]
Comment (dovics) [reply model/model.go:32]: Only `g` needs rolemanager, I think having one in the model is enough

File: enforcer.go [hunk 2]
```diff
@@ -26,20 +26,18 @@ import (
...
 	"github.com/casbin/casbin/v2/rbac"
-	defaultrolemanager "github.com/casbin/casbin/v2/rbac/default-role-manager"
 	"github.com/casbin/casbin/v2/util"
 )
 
 // Enforcer is the main interface for authorization enforcement and policy management.
 type Enforcer struct {
 	modelPath string
-	model     model.Model
+	model     *model.Model
```
Comment (nodece) [inline enforcer.go:35]: Why add `*` before `model.Model`?

**Comment by dovics** [comment]
@nodece @divy9881 The struct is passed as the value of the parameter, If we want to modify the value of model outside Casbin, the function parameters and return value must be pointer, for example
```
model := e.GetModel()
model.RM = defaultrolemanager.NewRoleManager(10) // This statement will not affect the model in the enforcer
```

**Comment by UnderTreeTech** [comment]
> Other than, `*model.Model` -> `model.Model`, it looks good to me.

Agree with the feature. It's better to use ptr than value. @nodece

**Comment by nodece** [comment]
thanks for @dovics answer, we have many breaking changes on recent PR, so I plan to use v3 and need to delay reviewing this PR.  

cc @GopherJ , @hsluoyz

**Comment by UnderTreeTech** [comment]
Any updates? @nodece

**Comment by hsluoyz** [comment]
Based on this comment: https://github.com/casbin/casbin/issues/501#issuecomment-653023905 , I don't think it's a very good idea to move role manager into model. I will close this for now. If the `remove Invalid field in AddMatchingFunc` part is still valuable, please raise another PR.

**Comment by divy9881** [comment]
> Based on this comment: #501 (comment) , I don't think it's a very good idea to move role manager into model. I will close this for now. If the remove Invalid field in AddMatchingFunc part is still valuable, please raise another PR.

I had already made the changes related to the irrelevant argument of MatchingFunc name 10 days back, in my branch when I was experimenting with the library, if its okay with @dovics then, I can make a PR?

**Pull request closed without merging**